libs/utils/src/main/java/com/akto/testing/HostValidator.java

```diff
@@ -0,0 +1,113 @@
+package com.akto.testing;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import com.akto.dto.OriginalHttpRequest;
+import com.akto.dto.testing.TestingRunConfig;
+import com.akto.dto.type.URLMethods;
+import com.akto.log.LoggerMaker;
+import com.akto.log.LoggerMaker.LogDb;
+import com.akto.util.DashboardMode;
+import okhttp3.Call;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+
+public class HostValidator {
+
+    public static final boolean SKIP_SSRF_CHECK = ("true".equalsIgnoreCase(System.getenv("SKIP_SSRF_CHECK")) || !DashboardMode.isSaasDeployment());
+    public static final boolean IS_SAAS = "true".equalsIgnoreCase(System.getenv("IS_SAAS"));
+
+    private static final LoggerMaker loggerMaker = new LoggerMaker(HostValidator.class, LogDb.TESTING);
+    
+    static Map<String, Boolean> hostReachabilityMap = new HashMap<>();
+
+    public static void validate(String url) throws Exception {
+        if (hostReachabilityMap != null) {
+
+            String checkUrl = getUniformUrl(url);
+
+            if (!hostReachabilityMap.getOrDefault(checkUrl, true)) {
+                loggerMaker.infoAndAddToDb(String.format(
+                        "Skipping url %s due to host unreachable previously", url));
+                throw new Exception("Host unreachable previously");
+            }
+        }
+    }
+
+    private static boolean checkDomainReach(Request request, boolean followRedirects, String requestProtocol) throws Exception{
+
+        if (HTTPClientHandler.instance == null) {
+            HTTPClientHandler.initHttpClientHandler(IS_SAAS);
+        }
+
+        OkHttpClient client = HTTPClientHandler.instance.getHTTPClient(followRedirects, requestProtocol);
+
+        if (!SKIP_SSRF_CHECK && !HostDNSLookup.isRequestValid(request.url().host())) {
+            throw new IllegalArgumentException("SSRF attack attempt");
+        }
+
+        Call call = client.newCall(request);
+        Response response = null;
+        try {
+            response = call.execute();
+        } catch (IOException e) {
+            if (!(request.url().toString().contains("insertRuntimeLog") || request.url().toString().contains("insertTestingLog") || request.url().toString().contains("insertProtectionLog"))) {
+                loggerMaker.errorAndAddToDb("Error while executing request " + request.url() + ": " + e, LogDb.TESTING);
+            } else {
+                System.out.println("Error while executing request " + request.url() + ": " + e);
+            }
+            return false;
+        } finally {
+            if (response != null) {
+                response.close();
+            }
+        }
+        return true;
+    }
+
+    private static String getUniformUrl(String url){
```

Comment:
write some unit tests for this:
1. with scheme, host and port
2. with scheme, host
3. only scheme